parquet-column/src/main/java/org/apache/parquet/internal/column/columnindex/ColumnIndexBuilder.java

```diff
@@ -293,8 +290,48 @@ boolean isNullPage(int pageIndex) {
 
     @Override
     public <T extends Comparable<T>> PrimitiveIterator.OfInt visit(In<T> in) {
-      IntSet indexes = getMatchingIndexes(in);
-      return IndexIterator.filter(getPageCount(), indexes::contains);
+      Set<T> values = in.getValues();
+      TreeSet<T> myTreeSet = new TreeSet<>();
+      IntSet matchingIndexes1 = new IntOpenHashSet();  // for null
+      Iterator<T> it = values.iterator();
+      while(it.hasNext()) {
+        T value = it.next();
+        if (value != null) {
+          myTreeSet.add(value);
+        } else {
+          if (nullCounts == null) {
+            // Searching for nulls so if we don't have null related statistics we have to return all pages
+            return IndexIterator.all(getPageCount());
+          } else {
+            for (int i = 0; i < nullCounts.length; i++) {
+              if (nullCounts[i] > 0) {
+                matchingIndexes1.add(i);
+              }
+            }
+          }
+        }
+      }
+
+      IntSet matchingIndexes2 = new IntOpenHashSet();
```

Comment:
I will think more about what is a better way to implement this `visit(In<T> in)`.

Reply:
@gszadovszky Sorry for the delay. I did the following changes:
1. Removed `TreeSet` to avoid sort the whole set, and added a method to get max and min.
2. Updated `StatisticsFilter` to have the similar range comparison.
3. I actually didn't change `notIn` because I don't think range checking works with `notIn`: if not in the range, `notIn` is true, but if in the range, it doesn't mean `notIn` is false.  For example, if we have `1, 2, 3, 6, 7 8, 9, 10` and the `notIn` predicate has set values `4, 5`, `4, 5` is in the range of 1 to 10, but `1, 2, 3, 6, 7 8, 9, 10` doesn't contain `4, 5`. In `StatisticsFilter` I simply return `BLOCK_MIGHT_MATCH;` for `notIn`. I probably should return `IndexIterator.all(getPageCount());` in `ColumnIndexBuilder` to be consistent with `StatisticsFilter`.

Reply:
The return types of `StatisticsFilter` and `ColumnIndexBuilder` are different because the first one simply returns whether the related row group can be dropped or not while the second one returns the indexes of the matching pages.

For `noIn` I agree. However there is an edge case you can figure out something when the min and max values are the same. In this case all the values shall be the same in the row group / page so you can say something in the case of there is only one element in `notIn`. It sounds a very rare scenario so I am fine returning "might match".